src/rfc9457/__init__.py

```diff
@@ -53,35 +58,36 @@ def __repr__(self: t.Self) -> str:
 
     @property
     def type(self: t.Self) -> str:
-        type_ = error_class_to_type(self)
-        return self._type if self._type else type_
+        return self._type if self._type else error_class_to_type(self)
 
-    def marshal(self: t.Self, type_base_url: str | None = None, *, strip_debug: bool = False) -> dict[str, t.Any]:
+    def marshal(
+        self: t.Self,
+        type_base_url: str | None = None,
+        *,
+        uri: str = "",
+        strip_debug: bool = False,
+    ) -> dict[str, t.Any]:
         """Generate a JSON compatible representation.
 
         Args:
         ----
-            type_base_url: If provided prepend to the type to generate a full url.
+            type_base_url: If provided prepend to the type to generate a full url. (DEPRECATED; use uri instead)
+            uri: URI / URI template to use as the type; will substitute '{status}', '{title}', and '{type}'
             strip_debug: If true, remove anything that is not title/type.
         """
-        type_ = f"{type_base_url or ''}{self.type}"
-        ret = {
-            "type": type_,
+        typ = self.type
+        if type_base_url
+            typ = f"{type_base_url or ''}{self.type}"
+            warn("Using deprecated parameter 'type_base_url'", DeprecationWarning)
+        typ = uri.replace("{status}", self.status).replace("{title}", self.title).replace("{type}", self.type) or self.type
```

Comment:
I personally do not use url types, so i have no real skin in this game, type_base_url was added to allow users to create a more rfc compatible response, this looks to improve on that and provide more configurability for users, im on board. feel free to update tests (good confirmation of maintained functionality) and tweak the docs.

Reply:
also regarding chained replaces, as long as we can asset uri is a string, using .format() wouldbr nicer
```
>>> "".format(status=self.status, type=self.type_, title=self.title)
''
>>>
```

Reply:
also regarding chained replaces, as long as we can assert uri is a string, using .format() would be nicer
```
>>> ''.format(status=self.status, type=self.type_, title=self.title)
''
>>>
```

Reply:
final line should also be `or typ` otherwise usage of type_base_url would get overwritten

Reply:
> using .format() would be nicer

Done -- I also added `**extras`.

Reply:
I also don't use URL `type`s. But, when I realized it was mandatory, I decided to just point to the Mozilla docs.

Reply:
The joy of the spec, is they don't have to resolve :D but they *can*
I just use type strings, to allow UI clients to make decisions based off a non moving string. If I had some massive project (like fastapi etc, then links to docs would be nice)

Reply:
Reading through the spec, it doesn't need to resolve if its a `tag:` URI. So i might expand on your work here to more closely match the spec.

Reply:
> it doesn't need to resolve if its a tag: URI

Oh, good catch!

For posterity, that's [stated in RFC 9457 section 3.1.1](https://www.rfc-editor.org/rfc/rfc9457.html#section-3.1.1-6):

> The type URI is allowed to be a non-resolvable URI. For example, the tag URI scheme [[TAG](https://www.rfc-editor.org/rfc/rfc9457.html#RFC4151)] can be used to uniquely identify problem types:
> `tag:example@example.org,2021-09-17:OutOfLuck`
> However, resolvable type URIs are encouraged by this specification because it might become desirable to resolve the URI in the future.